![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

# **Proyecto 1 - MDS7202 Laboratorio de Programación Científica para Ciencia de Datos 📚**

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### Cuerpo Docente:

- Profesor: Ignacio Meza, Gabriel Iturra
- Auxiliar: Sebastián Tinoco
- Ayudante: Arturo Lazcano, Angelo Muñoz

*Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir.*

### Equipo:

- Daniel Quilodrán
- Katherine Rutte


### Link de repositorio de GitHub: ``https://github.com/quilo98/Labs_MDS``

Fecha límite de entrega 📆: 27 de Octubre de 2023.

----

## Reglas

- **Grupos de 2 personas.**
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Estrictamente prohibida la copia. 
- Pueden usar cualquier material del curso que estimen conveniente.

<div style="text-align: center;">
    <img src="https://worldskateamerica.org/wp-content/uploads/2023/07/SANTIAGO-2023-1-768x153.jpg" alt="Descripción de la imagen">
</div>

En un Chile azotado por un profundo caos político-económico y el resurgimiento de programas de televisión de dudosa calidad, todas las miradas y esperanzas son depositadas en el éxito de un único evento: Santiago 2023. La nación necesitaba desesperadamente un respiro, y los Juegos de Santiago 2023 prometían ser una luz al final del túnel.

El Presidente de la República -conocido en las calles como Bombín-, consciente de la importancia de este evento para la revitalización del país, decide convocar a usted y su equipo en calidad de expertos en análisis de datos y estadísticas. Con gran solemnidad, el presidente les encomienda una importante y peligrosa: liderar un proyecto que permitiera caracterizar de forma automática y eficiente los datos generados por estos magnos juegos. Para esto, el presidente le destaca que la solución debe considerar los siguientes puntos:
- Caracterización automática de los datos
- La solución debe ser compatible con cualquier dataset
- Se les facilita el dataset *olimpiadas.parquet*, el cual recopila data de diferentes juegos olímpicos realizados en los últimos años

## 1.1 Creación de `Profiler` Class (4.0 puntos)

Cree la clase `Profiler`. Como mínimo, esta debe tener las siguientes funcionalidades:

1. El método constructor, el cual debe recibir los datos a procesar en formato `Pandas DataFrame`. Además, este método debe generar una carpeta en su directorio de trabajo con el nombre `EDA_fecha`, donde `fecha` corresponda a la fecha de ejecución en formato `DD-MM-YYYY`.

2. El método `summarize`, el cual debe caracterizar las variables del Dataset. Como mínimo, se espera que su método pueda:
    - Implementar una funcionalidad para filtrar y aplicar este método a una o más variables de interés.
    - Reportar el tipo de variable
    - Reportar el número y/o porcentaje de valores únicos de la variable
    - Reportar el número y/o porcentaje de valores nulos
    - Si la variables es numérica:
        - Reportar el número y/o porcentaje de valores cero, negativos y outliers
        - Reportar estadística descriptiva como el valor mínimo, máximo, promedio y los percentiles 25, 50, 75 y 100
   - Levantar una alerta en caso de encontrar alguna anomalía fuera de lo común (el criterio debe ser ajustable por el usuario)
   - Guardar sus resultados en el directorio `EDA_fecha/summary.txt`. El archivo debe separar de forma clara y ordenada los resultados de cada punto.

3. El método `plot_vars`, el cual debe graficar la distribución e interraciones de las variables del Dataset. Como mínimo, se espera que su método pueda:
    - Crear la carpeta `EDA_fecha/plots`
    - Implementar una funcionalidad para filtrar y aplicar este método a una o más variables de interés.
    - Para las variables numéricas:
        - Genere un gráfico de distribución de densidad
        - Grafique la correlación entre las variables
    - Para las variables categóricas:
        - Genere un histograma de las top N categorías (N debe ser un parámetro ajustable)
        - Grafique el coeficiente V de Cramer entre las variables
    - Guardar cada gráfico generado en la carpeta `EDA_fecha/plots` en formato `.pdf` y bajo el naming `variable.pdf`, donde `variable` es el nombre de la variable de interés
    
4. El método `clean_data`, el cual debe limpiar los datos para que luego puedan ser procesados. Como mínimo, se espera que su método pueda:
    - Crear la carpeta `EDA_fecha/clean_data`
    - Implementar una funcionalidad para filtrar y aplicar este método a una o más variables de interés.
    - Drop de valores duplicados
    - Implementar como mínimo 2 técnicas para tratar los valores nulos, como:
        - Drop de valores nulos
        - Imputar valores nulos con alguna técnica de imputación
        - Funcionalidad para escoger entre una técnica y la otra.
    - Una de las columnas del dataframe presenta datos *no atómicos*. Separe dicha columna en las columnas que la compongan.
        - Hint: ¿Qué caracteres permiten separar una columna de otra?
        - Para las pruebas con el dataset nuevo, puede esperar que exista al menos una columna con este tipo de problema. Asuma que los separadores serán los mismos, aunque el número de columnas a separar puede ser distinto.
    - Deberían usar `FunctionTransformer`.
    - Guardar los datos procesados en formato `.csv` en el path `EDA_fecha/clean_data/data.csv`

5. El método `scale`, el cual debe preparar adecuadamente los datos para luego ser consumidos por algún tipo de algoritmo. Como mínimo, se espera que su método pueda:
    - Crear la carpeta `EDA_fecha/scale`
    - Procesar de forma adecuada los datos numéricos y categóricos:
        - Su método debe recibir las técnicas de escalamiento como argumento de entrada (utilizar solo técnicas compatibles con el framework de `sklearn`)
        - Para los atributos numéricos, se transforme los datos con un escalador logarítmico y un `MinMaxScaler`
        - Asuma que no existen datos ordinales en su dataset
    - Guardar todo este procesamiento en un `ColumnTransformer`.
    - Guardar los datos limpios y transformados en formato `.csv` en el path `EDA_fecha/process/scaled_features.csv`

6. El método `make_clusters`, el cual debe generar clusters de los datos usando algún algoritmo de clusterización. Como mínimo, se espera que su método pueda:
    - Crear la carpeta `EDA_fecha/clusters`
    - Generar un estudio del codo donde señale la cantidad de clusters optimos para el desarrollo.
    - Su método debe recibir el algoritmo de clustering como argumento de entrada (utilizar solo algoritmos compatibles con el framework de `sklearn`).
    - No olvide pre procesar adecuadamente los datos antes de implementar la técnica de clustering. 
    - En este punto es espera que generen un `Pipeline` de sklearn. Además, su método debería usar lo construido en los puntos 4 y 5. 
    - Su método debe ser capaz de funcionar a partir de datos crudos (se descontará puntaje de lo contrario).
    - Una vez generado los clusters, proyecte los datos a 2 dimensiones usando su técnica de reducción de dimensionalidad favorita y grafique los resultados coloreando por cluster.
    - Guardar los datos con su respectivo cluster en formato `.csv` en el path `EDA_fecha/clusters/data_clusters.csv`. Guarde también los gráficos generados en el mismo path.

7. El método `detect_anomalies`, el cual debe detectar anomalías en los datos. Como mínimo, se espera que su método pueda:

    - Crear la carpeta `EDA_fecha/anomalies`
    - Implementar alguna técnica de detección de anomalías.
    - Al igual que el punto anterior, su método debe considerar los siguientes puntos:
        - No olvide pre procesar de forma adecuada los datos antes de implementar la técnica de detección de anomalía. 
        - En este punto es espera que generen un `Pipeline` de sklearn. Además, su método debería usar lo construido en los puntos 4 y 5. 
        - Su método debe ser capaz de funcionar a partir de datos crudos (se descontará puntaje de lo contrario).
        - Su método debe recibir el algoritmo como argumento de entrada
        - Una vez generado las etiquetas, proyecte los datos a 2 dimensiones y grafique los resultados coloreando por las etiquetas predichas por el detector de anomalías
    - Guardar los datos con su respectiva etiqueta en formato `.csv` en el path `EDA_fecha/anomalies/data_anomalies.csv`. Guarde también los gráficos generados en el mismo path.

8. El método `profile`, el cual debe ejecutar todos los métodos anteriores.

9. Crear el método `clearGarbage` para eliminar las carpetas/archivos creados/as por la clase `Profiler`.

Algunas consideraciones generales:
- Su clase será testeada con datos tabulares diferentes a los provistos. No desarrollen código *hardcodeado*: su clase debe ser capaz de funcionar para **cualquier** dataset. 
- Aplique todo su conocimiento sobre buenas prácticas de programación: se evaluará que su código sea limpio y ordenado.
- Recuerden documentar cada una de las funcionalidades que implementen.
- Recuerden adjuntar sus `requirements.txt` junto a su entrega de proyecto. **El código que no se pueda ejecutar por imcompatibilidades de librerías no será corregido.**

In [26]:
#Requirements
!pip install pyarrow

!pip install fastparquet

In [27]:
!pip install --upgrade numpy seaborn matplotlib

In [28]:
!pip install itertools

ERROR: Could not find a version that satisfies the requirement itertools (from versions: none)
ERROR: No matching distribution found for itertools


In [29]:
# Librerias necesarias 
import pandas as pd
import numpy as np
import pandas as pd
import datetime as dt
from IPython.display import HTML
#Libreria para plotear
import matplotlib.pyplot as plt
import os

import os
import seaborn as sns
from scipy import stats
import re
import itertools

from sklearn.compose import ColumnTransformer, TransformedTargetRegressor
from sklearn.preprocessing import FunctionTransformer, MinMaxScaler, OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.ensemble import IsolationForest
import shutil



In [30]:
# Visualizamos el dataset para ver con que variables nos podemos encontrar y asi tener una idea
df = pd.read_parquet('olimpiadas.parquet')
df.head()

ID                      Name Sex            Team  NOC        Games  Year  \
0   1                 A Dijiang   M           China  CHN  1992 Summer  1992   
1   2                  A Lamusi   M           China  CHN  2012 Summer  2012   
2   3       Gunnar Nielsen Aaby   M         Denmark  DEN  1920 Summer  1920   
3   4      Edgar Lindenau Aabye   M  Denmark/Sweden  DEN  1900 Summer  1900   
4   5  Christine Jacoba Aaftink   F     Netherlands  NED  1988 Winter  1988   

   Season       City          Sport                             Event Medal  \
0  Summer  Barcelona     Basketball       Basketball Men's Basketball  None   
1  Summer     London           Judo      Judo Men's Extra-Lightweight  None   
2  Summer  Antwerpen       Football           Football Men's Football  None   
3  Summer      Paris     Tug-Of-War       Tug-Of-War Men's Tug-Of-War  Gold   
4  Winter    Calgary  Speed Skating  Speed Skating Women's 500 metres  None   

  age-height-weight  
0   24.0*180.0?80.0  
1   23.0(170.0?60.0  
2      24.0(nan?nan  
3      34.0:nan?nan  
4   21.0(185.0?82.0

In [31]:
# Paleta de colores de los Juegos Panamericanos Santiago 2023

# Define los colores en formato hexadecimal
colores = ["#0074FF", "#FE4F3A", "#FFD200", "#00CCCC"]

# Crea la paleta personalizada
stgo2023 = sns.color_palette(colores)

# Configura Seaborn para utilizar la paleta personalizada
sns.set_palette(stgo2023)

In [32]:
class Profiler:
    def __init__(self, data_file):
        """
        Constructor de la clase Profiler.

        Parámetros:
        data_file (str): La ruta del archivo de datos que se desea analizar. 
        """
        self.data_file = data_file
        self.cleaned_data = None  # Añade un atributo para la data procesada
        self.scaled_data = None # Añade un atributo para la data escalada
        self.date = dt.datetime.now().strftime("%d-%m-%Y")
        self.output_dir = "EDA_{}".format(self.date)
        os.makedirs(self.output_dir, exist_ok=True)

        # Intenta cargar el archivo en un DataFrame de Pandas
        try:
            if data_file.lower().endswith('.csv'):
                self.data = pd.read_csv(data_file)
            elif data_file.lower().endswith(('.xls', '.xlsx')):
                self.data = pd.read_excel(data_file)
            elif data_file.lower().endswith('.parquet'):
                self.data = pd.read_parquet(data_file)
            else:
                raise ValueError("Formato de archivo no compatible. Solo se admiten archivos CSV, Excel y Parquet.")
        except Exception as e:
            raise Exception(f"No se pudo cargar el archivo: {str(e)}")

    def summarize(self, columns=None, anomaly_threshold=0.05):
        """
        Caracteriza las variables del DataFrame y guarda los resultados en un archivo summary.txt.

        Parámetros:
        columns (list): Lista de nombres de las columnas a analizar. Si es None, se analizarán todas las columnas.
        anomaly_threshold (float): Umbral para detectar anomalías en porcentaje. Por defecto, 0.05 (5%).

        Nota: anomaly_threshold es el porcentaje de valores únicos para considerar una variable como anómala.
        """
        summary_file = os.path.join(self.output_dir, "summary.txt")

        with open(summary_file, "w") as f:
            f.write(f"Resumen del Análisis - Fecha: {self.date}\n\n")

            if columns is None:
                columns = self.data.columns

            for col in columns:
                f.write(f"Variable: {col}\n")
                variable = self.data[col]
                variable_type = variable.dtype
                num_unique = variable.nunique()
                num_null = variable.isnull().sum()
                f.write(f"Tipo de Variable: {variable_type}\n")
                f.write(f"Número de Valores Únicos: {num_unique}\n")
                f.write(f"Número de Valores Nulos: {num_null}\n")

                if variable_type in ["int64", "float64"]:
                    num_zeros = (variable == 0).sum()
                    num_negatives = (variable < 0).sum()
                    num_duplicates = variable.duplicated().sum()
                    Q1 = variable.quantile(0.25)
                    Q3 = variable.quantile(0.75)
                    IQR = Q3 - Q1
                    lower_limit = Q1 - 1.5 * IQR
                    upper_limit = Q3 + 1.5 * IQR
                    num_outliers = ((variable < lower_limit) | (variable > upper_limit)).sum()
                    descriptive_stats = variable.describe()
                    f.write(f"Número de Valores Cero: {num_zeros}\n")
                    f.write(f"Número de Valores Negativos: {num_negatives}\n")
                    f.write(f"Número de Valores Duplicados: {num_duplicates}\n")
                    f.write(f"Número de Outliers: {num_outliers}\n")
                    f.write("Estadísticas Descriptivas:\n")
                    f.write(descriptive_stats.to_string() + "\n")

                    # Detectar anomalías basadas en el umbral de porcentaje
                    if (num_unique / len(variable)) > anomaly_threshold:
                        f.write("Anomalía: Esta variable tiene un alto porcentaje de valores únicos.\n")
                else:
                    if num_unique <= 5:
                        value_counts = variable.value_counts()
                        f.write("Frecuencia de Categorías:\n")
                        for category, count in value_counts.items():
                            f.write(f"{category}: {count}\n")
                    else:
                        top_categories = variable.value_counts().nlargest(5)
                        f.write("Frecuencia de las 5 Categorías Más Frecuentes:\n")
                        for category, count in top_categories.items():
                            f.write(f"{category}: {count}\n")
                    f.write("No se proporciona información adicional para variables no numéricas.\n")

                f.write("\n")


    def plot_vars(self, columns=None, top_n_categories=5):
        """
        Genera gráficos de distribución e interacciones de las variables y los guarda en la carpeta "plots".

        Parámetros:
        columns (list): Lista de nombres de las columnas a analizar. Si es None, se analizarán todas las columnas.
        top_n_categories (int): Número de categorías a mostrar en el histograma para variables categóricas.
        """
        plots_dir = os.path.join(self.output_dir, "plots")
        os.makedirs(plots_dir, exist_ok=True)

        if columns is None:
            columns = self.data.columns

        for col in columns:
            variable = self.data[col]
            sns.set_palette(stgo2023)
            plt.figure(figsize=(12, 8))

            if variable.dtype in ["int64", "float64"]:
                # Gráfico de distribución de densidad para variables numéricas
                sns.histplot(variable, kde=True)
                plt.title(f"Distribución de Densidad de {col}")
            else:
                # Histograma de las top N categorías para variables categóricas
                top_categories = variable.value_counts().head(top_n_categories)
                sns.barplot(x=top_categories.index, y=top_categories.values)
                plt.title(f"Top {top_n_categories} Categorías de {col}")
                plt.xticks(rotation=20)

            plt.xlabel(col)
            plt.ylabel("Frecuencia")
            plt.savefig(os.path.join(plots_dir, f"{col}.pdf"))
            plt.close()

        #Gráfico de correlación para variables numéricas
        num_vars = self.data.select_dtypes(include=["int64", "float64"])
        if not num_vars.empty:
            correlation_matrix = num_vars.corr()
            plt.figure(figsize=(12, 8))
            sns.heatmap(correlation_matrix, annot=True, cmap="coolwarm")
            plt.title("Correlación entre Variables Numéricas")
            plt.savefig(os.path.join(plots_dir, "correlation.pdf"))
            plt.close()

        #Gráfico del coeficiente V de Cramer entre variables categóricas
        cat_vars = self.data.select_dtypes(include=["object", "category"])
        if not cat_vars.empty:
            n = len(cat_vars.columns)
            plt.figure(figsize=(12, 8))
            cramer_matrix = np.ones((n, n))
            for i, col1 in enumerate(cat_vars.columns):
                for j, col2 in enumerate(cat_vars.columns):
                    if i != j:
                        try:
                            confusion_matrix = pd.crosstab(cat_vars[col1], cat_vars[col2])
                            chi2, _, _, _ = stats.chi2_contingency(confusion_matrix)
                            n = confusion_matrix.sum().sum()
                            cramer = np.sqrt(chi2 / (n * (min(confusion_matrix.shape) - 1)))
                            cramer_matrix[i, j] = cramer
                        except:
                            # Maneja la excepción (error) y continúa con las demás columnas
                            print(f"Error al calcular la matriz de confusión para {col1} y {col2}")
            sns.heatmap(cramer_matrix, annot=True, cmap="coolwarm", xticklabels=cat_vars.columns, yticklabels=cat_vars.columns)
            plt.title("Coeficiente V de Cramer entre Variables Categóricas")
            plt.savefig(os.path.join(plots_dir, "cramer.pdf"))
            plt.close()



    def clean_data(self, columns=None, drop_nulls=True, impute_strategy=None, impute_strategy_mapper=None):
        """
        Limpia los datos, eliminando valores duplicados, tratando valores nulos y separando columnas no atómicas.
        Guarda los datos procesados en formato .csv.

        Parámetros:
        columns (list): Lista de nombres de las columnas a las que se aplicará la limpieza.
                        Si es None, se aplicará a todas las columnas.
        drop_nulls (bool): Si es True, se eliminarán las filas con valores nulos. Si es False,
                        se imputarán los valores nulos con la técnica de imputación seleccionada.
        impute_strategy (str): Estrategia de imputación para valores nulos. Puede ser 'mean', 'median', 'most_frequent', o None.
        impute_strategy_mapper (dict): Un mapeo personalizado de estrategias de imputación por columna.
        """
        clean_dir = os.path.join(self.output_dir, "clean_data")
        os.makedirs(clean_dir, exist_ok=True)

        # Crear una copia del DataFrame original
        cleaned_data = self.data.copy()

        if columns is None:
            columns = cleaned_data.columns

        # Eliminar valores duplicados en la copia
        cleaned_data.drop_duplicates(subset=columns, inplace=True)

        # Definir caracteres especiales utilizando regex
        special_chars = r'[^A-Za-z0-9.]'
        columns_to_remove = []
        # Identificar columnas no atómicas en función de caracteres no alfanuméricos en los nombres de columna
        non_atomic_columns = [col for col in columns if re.search(special_chars, col)]

        # Separar columnas no atómicas
        for col in non_atomic_columns:
            if cleaned_data[col].dtype == "object":
                original_col_names = re.split(special_chars, col) 
                # Reemplazar caracteres especiales
                cleaned_data[col] = cleaned_data[col].str.replace(special_chars, ' ', regex=True)
                # Crear nuevas columnas a partir de los datos separados
                new_cols = cleaned_data[col].str.split(' ', expand=True)
                # Comprobar si el número de nuevas columnas es igual al de nombres extraídos, de lo contrario, puede ser un problema
                if new_cols.shape[1] != len(original_col_names):
                    print(f"Warning: The number of split parts in '{col}' does not match with expected names {original_col_names}")
                    # Puede querer manejar este caso con lógica adicional o aserciones aquí
                
                # Asignar los nombres de las columnas originales a las nuevas columnas
                new_cols.columns = original_col_names

                # Iterar a través de las nuevas columnas y convertirlas en numéricas o dejarlas como objetos
                for new_col in new_cols.columns:
                    new_cols[new_col] = new_cols[new_col].apply(lambda x: pd.to_numeric(x, errors='coerce'))

                # Añadir las nuevas columnas al DataFrame
                cleaned_data = pd.concat([cleaned_data, new_cols], axis=1)
                columns_to_remove.append(col)


        def custom_cleaning(X, drop_nulls, impute_strategy, impute_strategy_mapper):
            # Reemplazar 0, 'nan', 'NaN' con np.nan
            X.replace([0, 'nan', 'NaN'], np.nan, inplace=True)

            for col in X.columns:
                if X[col].dtype == 'object' or isinstance(X[col].dtype, pd.CategoricalDtype):
                    # Para columnas categóricas, reemplazar NaN con 'ND'
                    X[col].fillna('ND', inplace=True)
                else:
                    # Si la estrategia de imputación no es None, se imputan los valores numéricos como se especifica
                    if impute_strategy and (col not in X.select_dtypes(include=['object', 'category']).columns):
                        if impute_strategy_mapper and col in impute_strategy_mapper:
                            impute_strategy_col = impute_strategy_mapper[col]
                        else:
                            impute_strategy_col = impute_strategy
                        
                        if impute_strategy_col == 'mean':
                            X[col].fillna(X[col].mean(), inplace=True)
                        elif impute_strategy_col == 'median':
                            X[col].fillna(X[col].median(), inplace=True)
                        elif impute_strategy_col == 'most_frequent':
                            X[col].fillna(X[col].mode().iloc[0], inplace=True)
                        # Si se establece alguna otra estrategia de imputación, se coloca aquí
                    elif drop_nulls:
                        # Para columnas numéricas, si se elige eliminar valores nulos, se aplica aquí
                        X = X[pd.notnull(X[col])]

            return X


        # Crear el transformador con FunctionTransformer
        cleaning_transformer = FunctionTransformer(
            func=custom_cleaning,
            validate=False,
            kw_args={
                'drop_nulls': drop_nulls,
                'impute_strategy': impute_strategy,
                'impute_strategy_mapper': impute_strategy_mapper
            }
        )

        # Aplicar las transformaciones
        cleaned_data.drop(columns=columns_to_remove, inplace=True)
        cleaned_data = cleaning_transformer.transform(cleaned_data)
        processed_data_file = os.path.join(clean_dir, "data.csv")
        cleaned_data.to_csv(processed_data_file, index=False)
        self.cleaned_data = cleaned_data

    
    def scale(self, categorical_threshold=100):
        """
        Escala los datos utilizando una transformación logarítmica seguida de un Min-Max Scaler para variables numéricas
        y guarda el resultado en un archivo CSV.
        
        Parámetros:
        
        categorical_threshold (int): Umbral para codificar variables categóricas en one-hot.
        """
        scale_dir = os.path.join(self.output_dir, "scale")
        os.makedirs(scale_dir, exist_ok=True)

        # Si cleaned_data aún no está disponible, ejecuta clean_data
        if self.cleaned_data is None:
            self.clean_data()

        # Obtener las columnas numéricas y categóricas
        numerical_columns = self.cleaned_data.select_dtypes(include=["int64", "float64"]).columns
        categorical_columns = self.cleaned_data.select_dtypes(include=["object"]).columns

        # Identificar las columnas categóricas que tienen menos de `categorical_threshold` valores únicos
        categorical_columns_to_encode = [col for col in categorical_columns if self.cleaned_data[col].nunique() < categorical_threshold]

        # Crear un transformador secuencial para logaritmo y Min-Max Scaler para atributos numéricos
        numeric_transformer = Pipeline(steps=[
            ('log_transform', FunctionTransformer(func=np.log1p, validate=False)),
            ('minmax_scaler', MinMaxScaler())
        ])

        # Crear un transformador para atributos categóricos
        categorical_transformer = Pipeline(steps=[
            ('imputer', SimpleImputer(strategy='most_frequent')),
            ('onehot', OneHotEncoder())
        ])

        # Combinar los transformadores en un ColumnTransformer
        preprocessor = ColumnTransformer(
            transformers=[
                ('num', numeric_transformer, numerical_columns),
                ('cat', categorical_transformer, categorical_columns_to_encode)
            ])

        # Aplicar las transformaciones
        scaled_features = preprocessor.fit_transform(self.cleaned_data)
        
        # Obtener nombres de nuevas columnas después de one-hot encoding
        # Esto es necesario ya que el número de columnas aumenta después del one-hot encoding
        new_categorical_features = preprocessor.named_transformers_['cat']['onehot'].get_feature_names_out(categorical_columns_to_encode)

        # Combinar todos los nombres de columnas, numéricas y categóricas
        all_feature_names = numerical_columns.tolist() + list(new_categorical_features)

        # Crear un DataFrame con las características escaladas
        scaled_features = scaled_features.toarray()
        scaled_data = pd.DataFrame(data=scaled_features, columns=all_feature_names)
        
        # Guardar los datos procesados en formato .csv
        scaled_data_file = os.path.join(scale_dir, "scaled_features.csv")
        scaled_data.to_csv(scaled_data_file, index=False)
        self.scaled_data = scaled_data


    def make_clusters(self, cluster_algorithm=KMeans, n_clusters_range=range(2, 11)):
        """
        Genera clusters de los datos y guarda los resultados en la carpeta "clusters".

        Parámetros:
        cluster_algorithm (sklearn.cluster.ClusterMixin): Algoritmo de clustering de scikit-learn (KMeans) por defecto.
        n_clusters_range (range): Rango de valores para buscar el número óptimo de clusters.
        """
        clusters_dir = os.path.join(self.output_dir, "clusters")
        os.makedirs(clusters_dir, exist_ok=True)

        # Si cleaned_data aún no está disponible, ejecuta clean_data y scale
        if self.scaled_data is None:
            self.scale()
        cluster_data=self.scaled_data.copy()
        # Realizar el estudio del codo para encontrar el número óptimo de clusters en todas las dimensiones originales
        distortions = []
        for n_clusters in n_clusters_range:
            pipeline = Pipeline([
                ('cluster', cluster_algorithm(n_clusters=n_clusters, n_init=10))  # Algoritmo de clustering
            ])
            pipeline.fit(self.scaled_data)
            distortions.append(pipeline.named_steps['cluster'].inertia_)

        # Graficar el estudio del codo
        plt.figure(figsize=(12, 6))
        plt.plot(n_clusters_range, distortions, marker='o')
        plt.xlabel('Número de Clusters')
        plt.ylabel('Distorsión')
        plt.title('Método del Codo para selección del número de Clusters')
        plt.savefig(os.path.join(clusters_dir, 'Metodo_codo.pdf'))
        plt.close()
        # Elegir el número óptimo de clusters basado en el punto de quiebre en el estudio del codo
        optimal_n_clusters = n_clusters_range[np.argmin(np.diff(distortions)) + 1]
        print(optimal_n_clusters)
        # Realizar clustering con el número óptimo de clusters
        pipeline = Pipeline([
            ('cluster', cluster_algorithm(n_clusters=optimal_n_clusters, n_init=10))  # Algoritmo de clustering
        ])
        cluster_labels = pipeline.fit_predict(self.scaled_data)

        # Agregar la columna de cluster_labels al DataFrame original
        cluster_data['Cluster'] = cluster_labels
        cluster_data_file = os.path.join(clusters_dir, "data_clusters.csv")
        cluster_data.to_csv(cluster_data_file, index=False)


        # Proyectar los datos a 2 dimensiones
        pca = PCA(n_components=2)
        reduced_data = pca.fit_transform(self.scaled_data)
        reduced_data = pd.DataFrame(reduced_data, columns=['Dimension 1', 'Dimension 2'])


        # Graficar los resultados coloreando por cluster
        # Define los colores en formato hexadecimal
        colores = ["#0074FF", "#FE4F3A", "#FFD200", "#00CCCC"]
        # Crea la paleta personalizada
        stgo2023 = sns.color_palette(colores)
        plt.figure(figsize=(12, 8))
        sns.scatterplot(x='Dimension 1', y='Dimension 2', hue=cluster_data['Cluster'], data=reduced_data, palette=stgo2023)
        plt.title('Proyección de Datos en 2 Dimensiones con Coloreo por Cluster')
        plt.savefig(os.path.join(clusters_dir, 'cluster_plot.pdf'))

        # Guardar los gráficos generados
        plt.close()
        self.cluster_data=cluster_data


    def detect_anomalies(self, algoritmo=IsolationForest(contamination=0.05)):
        """
        Detecta anomalías en los datos y guarda los resultados en la carpeta "anomalies".

        Parámetros:
        algorithm (str): El algoritmo de detección de anomalías a utilizar. Por defecto tenemos 'isolation_forest'.
        """
        anomalies_dir = os.path.join(self.output_dir, "anomalies")
        os.makedirs(anomalies_dir, exist_ok=True)

        # Si los datos no están limpios y escalados, aplique los métodos clean_data y scale
        if self.cleaned_data is None:
            self.clean_data()
        if self.scaled_data is None:
            self.scale()


        anomaly_detection_pipeline = Pipeline([('algoritmo_anomalias', algoritmo)
        ])

        # Ajustar el modelo de detección de anomalías y obtener etiquetas
        anomaly_labels = anomaly_detection_pipeline.fit_predict(self.scaled_data)

        # Agregar las etiquetas de anomalías al DataFrame de datos
        anomalies_data = self.scaled_data.copy()
        anomalies_data['anomaly_labels'] = anomaly_labels

        # Guardar los datos con etiquetas de anomalías en formato .csv
        anomalies_data_file = os.path.join(anomalies_dir, "data_anomalies.csv")
        anomalies_data.to_csv(anomalies_data_file, index=False)

        # Reducir las dimensiones con PCA para graficar
        pca = PCA(n_components=2)
        reduced_data = pca.fit_transform(self.scaled_data)

        # Proyectar los datos a 2 dimensiones y graficar los resultados
        plt.figure(figsize=(8, 6))
        plt.scatter(reduced_data[:, 0], reduced_data[:, 1], c=anomaly_labels, cmap=plt.cm.Paired)
        plt.title("Detección de Anomalías")
        plt.xlabel("Dimensión 1")
        plt.ylabel("Dimensión 2")
        plt.colorbar()
        plt.savefig(os.path.join(anomalies_dir, "anomalies_plot.pdf"))
        plt.close()
        self.anomalies_data=anomalies_data





    def profile(self, columns=None, anomaly_threshold=0.05, top_n_categories=5, categorical_threshold=100, algoritmo=IsolationForest(contamination=0.05), cluster_algorithm=KMeans, n_clusters_range=range(2, 11)):
        """
        Ejecuta todos los métodos anteriores en orden.

        Parámetros:
        columns (list): Lista de nombres de las columnas a analizar en los métodos anteriores.
        anomaly_threshold (float): Umbral para detectar anomalías en porcentaje.
        top_n_categories (int): Número de categorías a mostrar en el histograma para variables categóricas.
        categorical_threshold (int): Umbral para codificar variables categóricas en one-hot.
        algorithm (str): El algoritmo de detección de anomalías a utilizar.
        cluster_algorithm (sklearn.cluster.ClusterMixin): Algoritmo de clustering de scikit-learn.
        n_clusters_range (range): Rango de valores para buscar el número óptimo de clusters.
        """
        self.clean_data(columns=columns)
        self.summarize(columns=columns, anomaly_threshold=anomaly_threshold)
        self.plot_vars(columns=columns, top_n_categories=top_n_categories)
        self.scale(categorical_threshold=categorical_threshold)
        self.make_clusters(cluster_algorithm=cluster_algorithm, n_clusters_range=n_clusters_range)
        self.detect_anomalies(algoritmo=algoritmo)


    
    def clearGarbage(self):
        """
        Elimina las carpetas y archivos creados por la clase Profiler.
        """
        if self.output_dir and os.path.exists(self.output_dir):
            shutil.rmtree(self.output_dir)  



In [33]:
perfil1 = Profiler('olimpiadas.parquet')
#perfil1.summarize()
#perfil1.plot_vars()
# perfil1.clean_data()
# perfil1.scale()
# perfil1.make_clusters()
perfil1.clearGarbage()
perfil1.profile()

c:\Users\quilo\anaconda3\envs\labs\lib\site-packages\pandas\core\reshape\reshape.py:143: RuntimeWarning: overflow encountered in scalar multiply
  num_cells = num_rows * num_columns


Error al calcular la matriz de confusión para Name y age-height-weight


c:\Users\quilo\anaconda3\envs\labs\lib\site-packages\pandas\core\reshape\reshape.py:143: RuntimeWarning: overflow encountered in scalar multiply
  num_cells = num_rows * num_columns


Error al calcular la matriz de confusión para age-height-weight y Name
3


C:\Users\quilo\AppData\Local\Temp\ipykernel_1644\3965703461.py:390: UserWarning: The palette list has more values (4) than needed (3), which may not be intended.
  sns.scatterplot(x='Dimension 1', y='Dimension 2', hue=cluster_data['Cluster'], data=reduced_data, palette=stgo2023)


In [34]:
perfil1.data

ID                      Name Sex            Team  NOC  \
0            1                 A Dijiang   M           China  CHN   
1            2                  A Lamusi   M           China  CHN   
2            3       Gunnar Nielsen Aaby   M         Denmark  DEN   
3            4      Edgar Lindenau Aabye   M  Denmark/Sweden  DEN   
4            5  Christine Jacoba Aaftink   F     Netherlands  NED   
...        ...                       ...  ..             ...  ...   
271111  135569                Andrzej ya   M        Poland-1  POL   
271112  135570                  Piotr ya   M          Poland  POL   
271113  135570                  Piotr ya   M          Poland  POL   
271114  135571        Tomasz Ireneusz ya   M          Poland  POL   
271115  135571        Tomasz Ireneusz ya   M          Poland  POL   

              Games  Year  Season            City          Sport  \
0       1992 Summer  1992  Summer       Barcelona     Basketball   
1       2012 Summer  2012  Summer          London           Judo   
2       1920 Summer  1920  Summer       Antwerpen       Football   
3       1900 Summer  1900  Summer           Paris     Tug-Of-War   
4       1988 Winter  1988  Winter         Calgary  Speed Skating   
...             ...   ...     ...             ...            ...   
271111  1976 Winter  1976  Winter       Innsbruck           Luge   
271112  2014 Winter  2014  Winter           Sochi    Ski Jumping   
271113  2014 Winter  2014  Winter           Sochi    Ski Jumping   
271114  1998 Winter  1998  Winter          Nagano      Bobsleigh   
271115  2002 Winter  2002  Winter  Salt Lake City      Bobsleigh   

                                           Event Medal age-height-weight  
0                    Basketball Men's Basketball  None   24.0*180.0?80.0  
1                   Judo Men's Extra-Lightweight  None   23.0(170.0?60.0  
2                        Football Men's Football  None      24.0(nan?nan  
3                    Tug-Of-War Men's Tug-Of-War  Gold      34.0:nan?nan  
4               Speed Skating Women's 500 metres  None   21.0(185.0?82.0  
...                                          ...   ...               ...  
271111                Luge Mixed (Men)'s Doubles  None   29.0:179.0?89.0  
271112  Ski Jumping Men's Large Hill, Individual  None   27.0:176.0?59.0  
271113        Ski Jumping Men's Large Hill, Team  None   27.0*176.0?59.0  
271114                      Bobsleigh Men's Four  None   30.0(185.0?96.0  
271115                      Bobsleigh Men's Four  None   34.0(185.0?96.0  

[271116 rows x 13 columns]

In [35]:
perfil1.cleaned_data

ID                      Name Sex         Team  NOC        Games  \
0            1                 A Dijiang   M        China  CHN  1992 Summer   
1            2                  A Lamusi   M        China  CHN  2012 Summer   
4            5  Christine Jacoba Aaftink   F  Netherlands  NED  1988 Winter   
5            5  Christine Jacoba Aaftink   F  Netherlands  NED  1988 Winter   
6            5  Christine Jacoba Aaftink   F  Netherlands  NED  1992 Winter   
...        ...                       ...  ..          ...  ...          ...   
271111  135569                Andrzej ya   M     Poland-1  POL  1976 Winter   
271112  135570                  Piotr ya   M       Poland  POL  2014 Winter   
271113  135570                  Piotr ya   M       Poland  POL  2014 Winter   
271114  135571        Tomasz Ireneusz ya   M       Poland  POL  1998 Winter   
271115  135571        Tomasz Ireneusz ya   M       Poland  POL  2002 Winter   

        Year  Season            City          Sport  \
0       1992  Summer       Barcelona     Basketball   
1       2012  Summer          London           Judo   
4       1988  Winter         Calgary  Speed Skating   
5       1988  Winter         Calgary  Speed Skating   
6       1992  Winter     Albertville  Speed Skating   
...      ...     ...             ...            ...   
271111  1976  Winter       Innsbruck           Luge   
271112  2014  Winter           Sochi    Ski Jumping   
271113  2014  Winter           Sochi    Ski Jumping   
271114  1998  Winter          Nagano      Bobsleigh   
271115  2002  Winter  Salt Lake City      Bobsleigh   

                                           Event Medal   age  height  weight  
0                    Basketball Men's Basketball    ND  24.0   180.0    80.0  
1                   Judo Men's Extra-Lightweight    ND  23.0   170.0    60.0  
4               Speed Skating Women's 500 metres    ND  21.0   185.0    82.0  
5             Speed Skating Women's 1,000 metres    ND  21.0   185.0    82.0  
6               Speed Skating Women's 500 metres    ND  25.0   185.0    82.0  
...                                          ...   ...   ...     ...     ...  
271111                Luge Mixed (Men)'s Doubles    ND  29.0   179.0    89.0  
271112  Ski Jumping Men's Large Hill, Individual    ND  27.0   176.0    59.0  
271113        Ski Jumping Men's Large Hill, Team    ND  27.0   176.0    59.0  
271114                      Bobsleigh Men's Four    ND  30.0   185.0    96.0  
271115                      Bobsleigh Men's Four    ND  34.0   185.0    96.0  

[206159 rows x 15 columns]

In [36]:
perfil1.scaled_data

ID      Year       age    height    weight  Sex_F  Sex_M  \
0       0.000000  0.804847  0.409636  0.604739  0.537908    0.0    1.0   
1       0.036449  0.967636  0.386853  0.505539  0.403674    0.0    1.0   
2       0.098760  0.772093  0.338291  0.652301  0.549454    1.0    0.0   
3       0.098760  0.772093  0.338291  0.652301  0.549454    1.0    0.0   
4       0.098760  0.804847  0.431525  0.652301  0.549454    1.0    0.0   
...          ...       ...       ...       ...       ...    ...    ...   
206154  0.999999  0.673435  0.511392  0.595069  0.587782    0.0    1.0   
206155  0.999999  0.983826  0.472886  0.565733  0.395849    0.0    1.0   
206156  0.999999  0.983826  0.472886  0.565733  0.395849    0.0    1.0   
206157  1.000000  0.853855  0.529692  0.652301  0.623237    0.0    1.0   
206158  1.000000  0.886445  0.597425  0.652301  0.623237    0.0    1.0   

        Games_1896 Summer  Games_1900 Summer  Games_1904 Summer  ...  \
0                     0.0                0.0                0.0  ...   
1                     0.0                0.0                0.0  ...   
2                     0.0                0.0                0.0  ...   
3                     0.0                0.0                0.0  ...   
4                     0.0                0.0                0.0  ...   
...                   ...                ...                ...  ...   
206154                0.0                0.0                0.0  ...   
206155                0.0                0.0                0.0  ...   
206156                0.0                0.0                0.0  ...   
206157                0.0                0.0                0.0  ...   
206158                0.0                0.0                0.0  ...   

        Sport_Triathlon  Sport_Tug-Of-War  Sport_Volleyball  Sport_Water Polo  \
0                   0.0               0.0               0.0               0.0   
1                   0.0               0.0               0.0               0.0   
2                   0.0               0.0               0.0               0.0   
3                   0.0               0.0               0.0               0.0   
4                   0.0               0.0               0.0               0.0   
...                 ...               ...               ...               ...   
206154              0.0               0.0               0.0               0.0   
206155              0.0               0.0               0.0               0.0   
206156              0.0               0.0               0.0               0.0   
206157              0.0               0.0               0.0               0.0   
206158              0.0               0.0               0.0               0.0   

        Sport_Weightlifting  Sport_Wrestling  Medal_Bronze  Medal_Gold  \
0                       0.0              0.0           0.0         0.0   
1                       0.0              0.0           0.0         0.0   
2                       0.0              0.0           0.0         0.0   
3                       0.0              0.0           0.0         0.0   
4                       0.0              0.0           0.0         0.0   
...                     ...              ...           ...         ...   
206154                  0.0              0.0           0.0         0.0   
206155                  0.0              0.0           0.0         0.0   
206156                  0.0              0.0           0.0         0.0   
206157                  0.0              0.0           0.0         0.0   
206158                  0.0              0.0           0.0         0.0   

        Medal_ND  Medal_Silver  
0            1.0           0.0  
1            1.0           0.0  
2            1.0           0.0  
3            1.0           0.0  
4            1.0           0.0  
...          ...           ...  
206154       1.0           0.0  
206155       1.0           0.0  
206156       1.0           0.0  
206157       1.0           0.0  
206158       1.0           0.0  

[

In [37]:
perfil1.cluster_data

ID      Year       age    height    weight  Sex_F  Sex_M  \
0       0.000000  0.804847  0.409636  0.604739  0.537908    0.0    1.0   
1       0.036449  0.967636  0.386853  0.505539  0.403674    0.0    1.0   
2       0.098760  0.772093  0.338291  0.652301  0.549454    1.0    0.0   
3       0.098760  0.772093  0.338291  0.652301  0.549454    1.0    0.0   
4       0.098760  0.804847  0.431525  0.652301  0.549454    1.0    0.0   
...          ...       ...       ...       ...       ...    ...    ...   
206154  0.999999  0.673435  0.511392  0.595069  0.587782    0.0    1.0   
206155  0.999999  0.983826  0.472886  0.565733  0.395849    0.0    1.0   
206156  0.999999  0.983826  0.472886  0.565733  0.395849    0.0    1.0   
206157  1.000000  0.853855  0.529692  0.652301  0.623237    0.0    1.0   
206158  1.000000  0.886445  0.597425  0.652301  0.623237    0.0    1.0   

        Games_1896 Summer  Games_1900 Summer  Games_1904 Summer  ...  \
0                     0.0                0.0                0.0  ...   
1                     0.0                0.0                0.0  ...   
2                     0.0                0.0                0.0  ...   
3                     0.0                0.0                0.0  ...   
4                     0.0                0.0                0.0  ...   
...                   ...                ...                ...  ...   
206154                0.0                0.0                0.0  ...   
206155                0.0                0.0                0.0  ...   
206156                0.0                0.0                0.0  ...   
206157                0.0                0.0                0.0  ...   
206158                0.0                0.0                0.0  ...   

        Sport_Tug-Of-War  Sport_Volleyball  Sport_Water Polo  \
0                    0.0               0.0               0.0   
1                    0.0               0.0               0.0   
2                    0.0               0.0               0.0   
3                    0.0               0.0               0.0   
4                    0.0               0.0               0.0   
...                  ...               ...               ...   
206154               0.0               0.0               0.0   
206155               0.0               0.0               0.0   
206156               0.0               0.0               0.0   
206157               0.0               0.0               0.0   
206158               0.0               0.0               0.0   

        Sport_Weightlifting  Sport_Wrestling  Medal_Bronze  Medal_Gold  \
0                       0.0              0.0           0.0         0.0   
1                       0.0              0.0           0.0         0.0   
2                       0.0              0.0           0.0         0.0   
3                       0.0              0.0           0.0         0.0   
4                       0.0              0.0           0.0         0.0   
...                     ...              ...           ...         ...   
206154                  0.0              0.0           0.0         0.0   
206155                  0.0              0.0           0.0         0.0   
206156                  0.0              0.0           0.0         0.0   
206157                  0.0              0.0           0.0         0.0   
206158                  0.0              0.0           0.0         0.0   

        Medal_ND  Medal_Silver  Cluster  
0            1.0           0.0        1  
1            1.0           0.0        1  
2            1.0           0.0        0  
3            1.0           0.0        0  
4            1.0           0.0        0  
...          ...           ...      ...  
206154       1.0           0.0        0  
206155       1.0           0.0        0  
206156       1.0           0.0        0  
206157       1.0           0.0        0  
206158       1.0           0.0        0  

[206159 rows x 163 columns]

In [38]:
import pandas as pd

# Agrupar los datos por el valor de la columna 'Cluster'
cluster_groups = perfil1.cluster_data.groupby('Cluster')

# Calcular estadísticas descriptivas para cada cluster
cluster_summaries = []
for cluster, cluster_data in cluster_groups:
    cluster_summary = cluster_data.describe()
    cluster_summary['Cluster'] = cluster  # Agregar el número de cluster como una columna
    cluster_summaries.append(cluster_summary)

# Concatenar los resúmenes de cada cluster en un solo DataFrame
cluster_summary_df = pd.concat(cluster_summaries)





In [39]:
display(cluster_summary_df[cluster_summary_df['Cluster']==0])

ID          Year           age        height        weight  \
count  39459.000000  39459.000000  39459.000000  39459.000000  39459.000000   
mean       0.916409      0.821971      0.425112      0.550036      0.473727   
std        0.084233      0.133434      0.097692      0.085984      0.080470   
min        0.098760      0.238879      0.000000      0.131298      0.112855   
25%        0.886385      0.739273      0.363100      0.485004      0.418945   
50%        0.938521      0.853855      0.431525      0.555844      0.475533   
75%        0.975958      0.918970      0.492471      0.614356      0.532027   
max        1.000000      0.983826      0.859739      0.839015      0.816793   

              Sex_F         Sex_M  Games_1896 Summer  Games_1900 Summer  \
count  39459.000000  39459.000000            39459.0            39459.0   
mean       0.337819      0.662181                0.0                0.0   
std        0.472973      0.472973                0.0                0.0   
min        0.000000      0.000000                0.0                0.0   
25%        0.000000      0.000000                0.0                0.0   
50%        0.000000      1.000000                0.0                0.0   
75%        1.000000      1.000000                0.0                0.0   
max        1.000000      1.000000                0.0                0.0   

       Games_1904 Summer  ...  Sport_Tug-Of-War  Sport_Volleyball  \
count            39459.0  ...           39459.0           39459.0   
mean                 0.0  ...               0.0               0.0   
std                  0.0  ...               0.0               0.0   
min                  0.0  ...               0.0               0.0   
25%                  0.0  ...               0.0               0.0   
50%                  0.0  ...               0.0               0.0   
75%                  0.0  ...               0.0               0.0   
max                  0.0  ...               0.0               0.0   

       Sport_Water Polo  Sport_Weightlifting  Sport_Wrestling  Medal_Bronze  \
count           39459.0              39459.0          39459.0  39459.000000   
mean                0.0                  0.0              0.0      0.040092   
std                 0.0                  0.0              0.0      0.196178   
min                 0.0                  0.0              0.0      0.000000   
25%                 0.0                  0.0              0.0      0.000000   
50%                 0.0                  0.0              0.0      0.000000   
75%                 0.0                  0.0              0.0      0.000000   
max                 0.0                  0.0              0.0      1.000000   

         Medal_Gold      Medal_ND  Medal_Silver  Cluster  
count  39459.000000  39459.000000  39459.000000        0  
mean       0.041182      0.877341      0.041385        0  
std        0.198713      0.328049      0.199181        0  
min        0.000000      0.000000      0.000000        0  
25%        0.000000      1.000000      0.000000        0  
50%        0.000000      1.000000      0.000000        0  
75%        0.000000      1.000000      0.000000        0  
max        1.000000      1.000000      1.000000        0  

[8 rows x 163 columns]

In [40]:

display(cluster_summary_df[cluster_summary_df['Cluster']==1])

ID           Year            age         height  \
count  113319.000000  113319.000000  113319.000000  113319.000000   
mean        0.909464       0.750415       0.435705       0.592592   
std         0.090608       0.178735       0.106744       0.096121   
min         0.000000       0.000000       0.044673       0.000000   
25%         0.874837       0.640416       0.363100       0.525835   
50%         0.937437       0.772093       0.431525       0.595069   
75%         0.973982       0.902715       0.492471       0.652301   
max         0.999997       1.000000       1.000000       1.000000   

              weight     Sex_F     Sex_M  Games_1896 Summer  \
count  113319.000000  113319.0  113319.0      113319.000000   
mean        0.504671       0.0       1.0           0.000353   
std         0.081463       0.0       0.0           0.018785   
min         0.051691       0.0       1.0           0.000000   
25%         0.448084       0.0       1.0           0.000000   
50%         0.501477       0.0       1.0           0.000000   
75%         0.555123       0.0       1.0           0.000000   
max         1.000000       0.0       1.0           1.000000   

       Games_1900 Summer  Games_1904 Summer  ...  Sport_Tug-Of-War  \
count      113319.000000      113319.000000  ...     113319.000000   
mean            0.000627           0.001050  ...          0.000176   
std             0.025023           0.032389  ...          0.013284   
min             0.000000           0.000000  ...          0.000000   
25%             0.000000           0.000000  ...          0.000000   
50%             0.000000           0.000000  ...          0.000000   
75%             0.000000           0.000000  ...          0.000000   
max             1.000000           1.000000  ...          1.000000   

       Sport_Volleyball  Sport_Water Polo  Sport_Weightlifting  \
count     113319.000000     113319.000000        113319.000000   
mean           0.015726          0.019688             0.022185   
std            0.124412          0.138926             0.147286   
min            0.000000          0.000000             0.000000   
25%            0.000000          0.000000             0.000000   
50%            0.000000          0.000000             0.000000   
75%            0.000000          0.000000             0.000000   
max            1.000000          1.000000             1.000000   

       Sport_Wrestling   Medal_Bronze     Medal_Gold       Medal_ND  \
count    113319.000000  113319.000000  113319.000000  113319.000000   
mean          0.043091       0.049824       0.050009       0.852840   
std           0.203062       0.217582       0.217965       0.354267   
min           0.000000       0.000000       0.000000       0.000000   
25%           0.000000       0.000000       0.000000       1.000000   
50%           0.000000       0.000000       0.000000       1.000000   
75%           0.000000       0.000000       0.000000       1.000000   
max           1.000000       1.000000       1.000000       1.000000   

        Medal_Silver  Cluster  
count  113319.000000        1  
mean        0.047327        1  
std         0.212337        1  
min         0.000000        1  
25%         0.000000        1  
50%         0.000000        1  
75%         0.000000        1  
max         1.000000        1  

[8 rows x 163 columns]

In [41]:

display(cluster_summary_df[cluster_summary_df['Cluster']==2])

ID          Year           age        height        weight  \
count  53381.000000  53381.000000  53381.000000  53381.000000  53381.000000   
mean       0.915308      0.831052      0.387470      0.484245      0.397148   
std        0.083799      0.139142      0.125859      0.096612      0.082605   
min        0.174927      0.204967      0.000000      0.000000      0.000000   
25%        0.880738      0.739273      0.285097      0.421909      0.345975   
50%        0.941196      0.870158      0.386853      0.485004      0.395849   
75%        0.975583      0.935208      0.472886      0.545898      0.448084   
max        0.999998      1.000000      0.984278      0.897064      0.883234   

         Sex_F    Sex_M  Games_1896 Summer  Games_1900 Summer  \
count  53381.0  53381.0            53381.0            53381.0   
mean       1.0      0.0                0.0                0.0   
std        0.0      0.0                0.0                0.0   
min        1.0      0.0                0.0                0.0   
25%        1.0      0.0                0.0                0.0   
50%        1.0      0.0                0.0                0.0   
75%        1.0      0.0                0.0                0.0   
max        1.0      0.0                0.0                0.0   

       Games_1904 Summer  ...  Sport_Tug-Of-War  Sport_Volleyball  \
count            53381.0  ...           53381.0      53381.000000   
mean                 0.0  ...               0.0          0.028006   
std                  0.0  ...               0.0          0.164992   
min                  0.0  ...               0.0          0.000000   
25%                  0.0  ...               0.0          0.000000   
50%                  0.0  ...               0.0          0.000000   
75%                  0.0  ...               0.0          0.000000   
max                  0.0  ...               0.0          1.000000   

       Sport_Water Polo  Sport_Weightlifting  Sport_Wrestling  Medal_Bronze  \
count      53381.000000         53381.000000     53381.000000  53381.000000   
mean           0.009142             0.008617         0.005676      0.054701   
std            0.095176             0.092429         0.075127      0.227398   
min            0.000000             0.000000         0.000000      0.000000   
25%            0.000000             0.000000         0.000000      0.000000   
50%            0.000000             0.000000         0.000000      0.000000   
75%            0.000000             0.000000         0.000000      0.000000   
max            1.000000             1.000000         1.000000      1.000000   

         Medal_Gold      Medal_ND  Medal_Silver  Cluster  
count  53381.000000  53381.000000  53381.000000        2  
mean       0.053858      0.837676      0.053764        2  
std        0.225740      0.368751      0.225554        2  
min        0.000000      0.000000      0.000000        2  
25%        0.000000      1.000000      0.000000        2  
50%        0.000000      1.000000      0.000000        2  
75%        0.000000      1.000000      0.000000        2  
max        1.000000      1.000000      1.000000        2  

[8 rows x 163 columns]

In [42]:
perfil1.anomalies_data

ID      Year       age    height    weight  Sex_F  Sex_M  \
0       0.000000  0.804847  0.409636  0.604739  0.537908    0.0    1.0   
1       0.036449  0.967636  0.386853  0.505539  0.403674    0.0    1.0   
2       0.098760  0.772093  0.338291  0.652301  0.549454    1.0    0.0   
3       0.098760  0.772093  0.338291  0.652301  0.549454    1.0    0.0   
4       0.098760  0.804847  0.431525  0.652301  0.549454    1.0    0.0   
...          ...       ...       ...       ...       ...    ...    ...   
206154  0.999999  0.673435  0.511392  0.595069  0.587782    0.0    1.0   
206155  0.999999  0.983826  0.472886  0.565733  0.395849    0.0    1.0   
206156  0.999999  0.983826  0.472886  0.565733  0.395849    0.0    1.0   
206157  1.000000  0.853855  0.529692  0.652301  0.623237    0.0    1.0   
206158  1.000000  0.886445  0.597425  0.652301  0.623237    0.0    1.0   

        Games_1896 Summer  Games_1900 Summer  Games_1904 Summer  ...  \
0                     0.0                0.0                0.0  ...   
1                     0.0                0.0                0.0  ...   
2                     0.0                0.0                0.0  ...   
3                     0.0                0.0                0.0  ...   
4                     0.0                0.0                0.0  ...   
...                   ...                ...                ...  ...   
206154                0.0                0.0                0.0  ...   
206155                0.0                0.0                0.0  ...   
206156                0.0                0.0                0.0  ...   
206157                0.0                0.0                0.0  ...   
206158                0.0                0.0                0.0  ...   

        Sport_Tug-Of-War  Sport_Volleyball  Sport_Water Polo  \
0                    0.0               0.0               0.0   
1                    0.0               0.0               0.0   
2                    0.0               0.0               0.0   
3                    0.0               0.0               0.0   
4                    0.0               0.0               0.0   
...                  ...               ...               ...   
206154               0.0               0.0               0.0   
206155               0.0               0.0               0.0   
206156               0.0               0.0               0.0   
206157               0.0               0.0               0.0   
206158               0.0               0.0               0.0   

        Sport_Weightlifting  Sport_Wrestling  Medal_Bronze  Medal_Gold  \
0                       0.0              0.0           0.0         0.0   
1                       0.0              0.0           0.0         0.0   
2                       0.0              0.0           0.0         0.0   
3                       0.0              0.0           0.0         0.0   
4                       0.0              0.0           0.0         0.0   
...                     ...              ...           ...         ...   
206154                  0.0              0.0           0.0         0.0   
206155                  0.0              0.0           0.0         0.0   
206156                  0.0              0.0           0.0         0.0   
206157                  0.0              0.0           0.0         0.0   
206158                  0.0              0.0           0.0         0.0   

        Medal_ND  Medal_Silver  anomaly_labels  
0            1.0           0.0               1  
1            1.0           0.0               1  
2            1.0           0.0               1  
3            1.0           0.0               1  
4            1.0           0.0              -1  
...          ...           ...             ...  
206154       1.0           0.0               1  
206155       1.0           0.0              -1  
206156       1.0           0.0              -1  
206157       1.0           0.0               1  
206158       1.0           0.0               1  

[206159 rows 

In [43]:
import pandas as pd

def summary_of_anomalies(dataframe):
    # Filtrar las instancias normales y anómalas
    normal_instances = dataframe[dataframe['anomaly_labels'] == 1]
    anomaly_instances = dataframe[dataframe['anomaly_labels'] == -1]

    # Resumen de las instancias normales
    normal_summary = normal_instances.describe()

    # Resumen de las instancias anómalas
    anomaly_summary = anomaly_instances.describe()

    return normal_summary, anomaly_summary

# Llamar a la función con tu DataFrame
normal_summary, anomaly_summary = summary_of_anomalies(perfil1.anomalies_data)

# Mostrar los resúmenes
print("Resumen de instancias normales:")
print(normal_summary)

print("\nResumen de instancias anómalas:")
print(anomaly_summary)

Resumen de instancias normales:
                  ID           Year            age         height  \
count  195851.000000  195851.000000  195851.000000  195851.000000   
mean        0.912795       0.786077       0.421335       0.556910   
std         0.086692       0.166117       0.111745       0.102478   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.879139       0.673435       0.338291       0.485004   
50%         0.938959       0.821199       0.409636       0.555844   
75%         0.974578       0.918970       0.492471       0.633432   
max         1.000000       1.000000       1.000000       1.000000   

              weight          Sex_F          Sex_M  Games_1896 Summer  \
count  195851.000000  195851.000000  195851.000000      195851.000000   
mean        0.471180       0.318089       0.681911           0.000204   
std         0.091316       0.465735       0.465735           0.014290   
min         0.051691       0.000000       0.000000    

## 1.2 Caracterizar datos de Olimpiadas (2.0 puntos)

A partir de la clase que hemos desarrollado previamente, procederemos a realizar un análisis exhaustivo de los datos proporcionados en el enunciado. Este análisis se presentará en forma de un informe contenido en el mismo Jupyter Notebook y abordará los siguientes puntos:

1. Introducción
    - Se proporcionará una breve descripción del problema que estamos abordando y se explicará la metodología que se seguirá.

Elaborar una breve introducción con todo lo necesario para entender qué realizarán durante su proyecto. La idea es que describan de manera formal el proyecto con sus propias palabras y logren describir algunos aspectos básicos tanto del dataset como del análisis a realizar sobre los datos.

Por lo anterior, en esta sección ustedes deberán ser capaces de:

- Describir la tarea asociada al dataset.
- Describir brevemente los datos de entrada que les provee el problema.
- Plantear hipótesis de cómo podrían abordar el problema.

2. Análisis del EDA (Análisis Exploratorio de Datos)
    - Se discutirán las observaciones y conclusiones obtenidas acerca de los datos proporcionados. A lo largo de su respuesta, debe responder preguntas como:
        - ¿Como se comportan las variables numéricas? ¿y las categóricas?
        - ¿Existen valores nulos en el dataset? ¿En qué columnas? ¿Cuantos?
        - ¿Cuáles son las categorías y frecuencias de las variables categóricas?
        - ¿Existen datos duplicados en el conjunto?
        - ¿Existen relaciones o patrones visuales entre las variables?
        - ¿Existen anomalías notables o preocupantes en los datos?
3. Creación de Clusters y Anomalías
    - Se justificará la elección de los algoritmos a utilizar y sus hiperparámetros. En el caso de clustering, justifique además el número de clusters.
    
4. Análisis de Resultados
    - Se examinarán los resultados obtenidos a partir de los clústers y anomalías generadas. ¿Se logra una separación efectiva de los datos? Entregue una interpretación de lo que representa cada clúster y anomalía.
5. Conclusión
    - Se resumirán las principales conclusiones del análisis y se destacarán las implicaciones prácticas de los resultados obtenidos.

1. La tarea desarrollada se centra en poder caracterizar de forma automática los datos obtenidos de los Juegos de Santiago 2023, para esto se solicita desarrollar una clase que pueda cumplir con dicha tarea para cualquier tipo de dataset, la clase en particular debe ser capaz de generar un resumen de los datos obtenidos, así como aplicar una limpieza de los datos  y un escalamiento de estos para su posterior uso, además se espera generar una cantidad óptima de Clusters para segmentar a los distintos atletas que participen en dichos juegos y detectar anomalías dentro de los datos. Para la creación de esta clase y a modo de ejemplo se utiliza un dataset llamado Olimpiadas que contiene información de los distintos atletas que han participado en los distintos juegos olimpicos que se han realizado desde el año 1896 hasta el 2016, este periodo abarca 51 de estos juegos, la información de los datos abarca el nombre, país, año de competencia, juego olímpico en el que participó, deporte, medalla ganada, entre otros.

2. El dataset cuenta con 2 variables numéricas: 'ID' y 'Year', Además cuenta con 11 variables categóricas: 'Name', 'Sex', 'Team', 'NOC', 'Games', 'Season', 'City', 'Sport', 'Event', 'Medal', 'age-height-weight'. a continuación se describe cada una de estas variables, para las variables categóricas también se muestran la frecuencia de las categorías, en caso de poseer mas de 5 categorías se presentan las 5 con mayor frecuencia:
   - ID: Presenta un ID único para cada atleta
   - Year: Presenta el año en el que fueron realizados los juegos
   - Name: Corresponde al nombre del atleta
      - Robert Tait McKenzie: 58
      - Heikki Ilmari Savolainen: 39
      - Joseph "Josy" Stoffel: 38
      - Ioannis Theofilakis: 36
      - Takashi Ono: 33
   - Sex: Corresponde al genero del atleta (F o M)
      - M: 196.594
      - F: 74.522
   - Team: Equipo al que pertenece el atleta (en general es el país por el que participa)
      - United States: 17.847
      - France: 11.988
      - Great Britain: 11.404
      - Italy: 10.260
      - Germany: 9.326
   - NOC: Abreviatura de 3 letras para el Team
      - USA: 18.853
      - FRA: 12.758
      - GBR: 12.256
      - ITA: 10.715
      - GER: 9.830
   - Games: Año y nombre del juego olímpico
      - 2000 Summer: 13.821
      - 1996 Summer: 13.780
      - 2016 Summer: 13.688
      - 2008 Summer: 13.602
      - 2004 Summer: 13.443
   - Season: Estación del año en el que se realizaron los juegos
      - Summer: 222.552
      - Winter: 48.564
   - City: Ciudad donde se realizó el evento deportivo en el que participó el atleta
      - London: 22.426
      - Athina: 15.556
      - Sydney: 13.821
      - Atlanta: 13.780
      - Rio de Janeiro: 13.688
   - Sport: Deporte en el que participó el atleta
      - Athletics: 38.624
      - Gymnastics: 26.707
      - Swimming: 23.195
      - Shooting: 11.448
      - Cycling: 10.859
   - Event: Evento en el que participó el atleta, especifica la modalidad del deporte de la competencia
      - Football Men's Football: 5.733
      - Ice Hockey Men's Ice Hockey: 4.762
      - Hockey Men's Hockey: 3.958
      - Water Polo Men's Water Polo: 3.358
      - Basketball Men's Basketball: 3.280
   - Medal: Presenta la medalla ganada por el atleta (Gold, Silver y Bronze), en caso de no haber ganado una medalla se presenta con NA
      - Gold: 13.372
      - Bronze: 13.295
      - Silver: 13.116
   - age-height-weight: Variable no atómica que presenta la edad, la altura y el peso del atleta

   En cuanto a los valores nulos estos solo se presentan en la variable 'Medal', esta variable cuenta con 231.333 valores nulos, lo que señala que la mayoría de atletas no logró ganar una medalla. las variables categóricas presentan un rango significativo en cuanto a la cantidad de categorías que presentan, siendo 'sex' la variable con menos categorías presentando solo 2 (M y F), mientras que la variable 'Name' es la que presenta una mayor cantidad de categorías con 134.732 (Cada nombre de atleta es una categoría). Se presentan datos duplicados en ambas variables númericas, en caso de ID presenta 135.545 datos duplicados y Year presenta 271.081, esto tiene sentido ya que el ID está asociado a cada atleta y los atletas generalmente participan en mas de un juego olímpico y en mas de una competencia, además la columna Year repetirá su valor por cada competencia y atleta que participó en esta en un juego en particular. En cuanto a las relaciones entre variables, para las númericas se tiene una correlación de 0.012 entre ID y Year, lo cual no presenta una correlación significativa. Para las variables categóricas se utilizó el coeficiente V de Cramer para analizar las relaciones, los resultados de estas muestran una relación significativa entre variables que identifican al atleta como Name y age-height-weight, y en las variables que identifican los juegos como: Game, City y season. Luego una de las relaciones interesantes que se observan es la relación entre Medals y age-height-weight que presenta un alto coeficiente de Cramer, por otro lado la variable Medal no presenta alguna otra relación relevante. Por último la principal anomalía observada es en la variable age-height-weight tanto por ser una variable de datos no atómicos y porque se presentan bastantes na para alguna de las 3 variables que representa. 


3. A continuación se justifica el uso de los algoritmos para la creación de cluster y detección de anomalías:

   * Clustering: Se utiliza el algoritmo Kmeans como predeterminado ya que es uno de los algoritmos de clustering más utilizados debido a su eficiencia y simplicidad, es un buen primer acercamiento para la creación de clusters en comparación con algoritmos mas complejos y que necesitan una mayor comprensión previa de los datos. En cuanto a sus hiperparámetros se establece el valor de n_init en 10, esto implica que el algoritmo se ejecuta 10 veces con diferentes centroides iniciales y se selecciona la mejor asignación de clusters para así reducir la sensibilidad del resultado frente a la inicialización aleeatoria. Por último para la elección del número óptimo de Clusters se utiliza el método del codo, este metodo utiliza como métrica la distorción para seleccionar el número óptimo de Clusters. La distorsión es una medida de que tan cerca están los puntos dentro de cada cluster respecto a su centroide, mientras menor sea la distorción mejor es la agrupación de puntos en los Clusters, la distorción tiende a disminuir según se aumenta la cantidad de Clusters. El método del codo utiliza esto último para encontrar el punto donde la distorción presenta el mínimo cambio al aumentar la cantidad de Clusters, en este caso en particular el número óptimo de Clusters es 3.

   * Detección de anomalías: Se utiliza el algorimo Isolation Forest como algoritmo determinado, este algoritmo de detección de anomalías se basa en la idea de que las anomalías son puntos atípicos que son más fáciles de aislar en comparación con los puntos "normales" en un conjunto de datos, se escoge este algoritmo ya que se destaca por su robustez frente a valores atípicos, su eficiencia para trabajar con grandes cantidades de datos y su facilidad de usar al solo utilizar un hiperparámetro, este hiperparámetro es 'contamination', contamination controla la proporción estimada de anomalías en el conjunto de datos. Se establece como una fracción del total de puntos de datos y representa la cantidad de puntos de datos que se espera que sean anomalías. Se utiliza un valor de 0.05 para este hiperparámetro, que es un valor típico para este algoritmo en particular, este valor significa que se asume que alrededor del 5% de los datos son anomalías.


4. Al analizar los clusters algo que se observa a simple vista es que el Cluster N° 1 tiene una relación de 66% atletas hombre y 34% mujeres en el cluster, mientras que el Cluster N° 2 presenta solo hombres y el Cluster N° 3 presenta solo mujeres. En los 3 Clusters la distibución de medallas es similar, manteniendo una distribución cercana al 85% para los atletas sin medalla, un 5% para atletas de medalla de bronce, plata u oro, exceptuando el Cluster N° 1 que presenta un 4% de atletas con medalla de bronce, plata u oro. Los valores de peso y altura en promedio son mayores para el Cluster de solo hombres, seguido por el Cluster mixto y finalmente el Cluster de mujeres. En cuanto a la edad el cluster de mujeres presenta una edad promedio menor, mientras que los clusters de hombres y mixtos presentan un valor mayor, los 3 cluster presentan una edad promedio menor a la media de los datos escalados. 
   
Luego al analizar las anomalías detectadas, el algoritmo detectó 10,308 anomalías y 195,851 datos que no son anomalías, se observan diferencias en las medias y desviaciones estandar de las variables age, height, weight y year, además de que en las anomalías se observa una distribución mucho mas equitativa en cuanto a los ateltas y si ganaron o no medallas en comparación a los datos no anómalos donde los atletas que no ganaron medallas representan un poco mas del 80% de los datos, por último las participaciones en distintos deportes y juegos también muestra una diferencia en cuanto a las proporciones entre ambos datos.

5. En la realización de la tarea en cuestión se obtiene información detallada de los juegos a analizar, obteniendo un resumen de los datos y poder segmentar a los atletas que participaron en los distintos juegos, se obtienen 3 clusters claramente separados al utilizar el algoritmo PCA para poder observarlo en 2 dimensiones, la segmentación de los atletas permite poder realizar un análisis mas profundo a cada cluster y los atletas que los componen, además la detección de anomalías permite detectar nuevas maneras de mejorar los datos y la obtención de estos en un futuro.